File: javascript/easy/binaryLevelOrderTraversalII.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function(root) {
    if(root == null) return [];
    let queue = [], res = [], level = [], numNodesInLevel = 0, currentNodes = 0;
    queue.unshift(root);
    numNodesInLevel = queue.length;
    while(queue.length > 0) {
        let current = queue.pop();
        currentNodes++;
        level.push(current.val);

        if(current.left != null) {
            queue.unshift(current.left);
        }

        if(current.right != null) {
            queue.unshift(current.right);
        }

        if(currentNodes == numNodesInLevel) {
            res.unshift(level);
            level = [];
            numNodesInLevel = queue.length;
            currentNodes = 0;
        }
    }
    return res;
};

File: ruby/easy/binary_level_order_traversal_ii.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer[][]}
def level_order_bottom(root)
    return [] unless root
    queue, result, level = [], [], []
    num_nodes, current_count = 1, 0
    queue.unshift(root)
    while(!queue.empty?)
        current = queue.pop()
        current_count += 1
        level.push(current.val)

        queue.unshift(current.left) if current.left
        queue.unshift(current.right) if current.right

        if(current_count == num_nodes)
            result.unshift(level)
            level = []
            current_count = 0
            num_nodes = queue.length
        end
    end
    return result
end
